### ISSUE TITLE
nixos-generate-config request: put in useDHCP stuff in hardware-configuration.nix

### ISSUE THREAD

**Issue by jamescostian** [NixOS/nixpkgs#73595]
As I understand it, `hardware-configuration.nix` contains hardware-specific details, while `configuration.nix` is made to be more easily shared. But `nixos-generate-config` puts things like `networking.interfaces.enp0s31f6.useDHCP = true;` in `configuration.nix` on some machines, and puts things like ` networking.interfaces.wlo1.useDHCP = true;` in `configuration.nix` on other computers.

When installing on a machine, one may try to use `nixos-generate-config` on it, and then copy a `configuration.nix` from another computer - but that won't put the correct `useDHCP` incantations in unless you have the same network interfaces on both machines!

**Comment by davidak** [comment]
When you put `networking.interfaces.enp0s31f6.useDHCP = true;` into `hardware-configuration.nix`, you can't change it to `false` without it getting overwritten by the next run of `nixos-generate-config`. (you might want to do that when you added a new harddrive or changed other hardware)

You also have other hardware specific settings in `configuration.nix`, like `boot.loader.grub.device = "/dev/sda";` (would be xvda on a Xen VM) or your static IP config.

I recommend to use a modular config when you have multiple NixOS systems. See my config for example. I have a `configuration.nix` for every machine, but have shared configs in  "profiles", like desktop or gaming.

https://github.com/davidak/nixos-config/blob/504cc5c256c337486b6598ba07e46b2307f0d759/machines/thinkpad_X230/configuration.nix

This is not a bug or regression and should not block 20.03 release.

**Comment by bjornfor** [comment]
> When you put networking.interfaces.enp0s31f6.useDHCP = true; into hardware-configuration.nix, you can't change it to false without it getting overwritten by the next run of nixos-generate-config. (you might want to do that when you added a new harddrive or changed other hardware)

Good point. But I think that just means the setting should be generated with mkDefault in hardware-configuration.nix and we should assume that the user will override this (permanently) in configuration.nix, _if they like_.

> You also have other hardware specific settings in configuration.nix, like boot.loader.grub.device = "/dev/sda"; (would be xvda on a Xen VM) or your static IP config.

"/dev/sda" can be removed if using UEFI. Having static IP in configuration.nix is something the user may chose to do, it's not forced by the installer.

> This is not a bug or regression

I disagree. Creating auto-install script for NixOS becomes harder now, because one cannot deploy a fixed configuration.nix.

> and should not block 20.03 release.

I realize it will not (based on feedback in https://github.com/NixOS/nixpkgs/issues/75515), but I wish it did.

I think this issue is a result of deprecating networking.useDHCP, which I object to: https://github.com/NixOS/nixpkgs/issues/75515.

**Comment by davidak** [comment]
>I think this issue is a result of deprecating networking.useDHCP, which I object to: #75515.

yes, that is my point. this specific problem should be fixed in that issue

i  understand this issue as there shouldn't be system specific confic in `configuration.nix` but i think they do belong there.

defaults should be in nixos module and not in `hardware-configuration.nix`

**Comment by bjornfor** [comment]
>i understand this issue as there shouldn't be system specific confic in configuration.nix but i think they do belong there.
>
> defaults should be in nixos module and not in hardware-configuration.nix

But the network names are hardware specific and cannot be shared between machines. nixos-generate-config literally inspects the current hardware. I really don't understand the benefit of having that in configuration.nix. I only see downsides:

* configuration.nix cannot be shared and by extension one cannot create a fully automatic NixOS installer (if you want to have DHCP enabled, thanks to https://github.com/NixOS/nixpkgs/issues/75515 being a thing).
* adding or removing a NIC means (1) running nixos-generate-config only to see it updates hardware-configuration.nix and not configuration.nix, then having to (2) manually copy over the entries.

I find deprecating `networking.useDHCP` a terribly bad idea, but at least if we fix _this_ issue, by moving the generated stuff into hardware-configuration.nix and use mkDefault, one can still share configuration.nix in NixOS 20.03+.

**Comment by davidak** [comment]
>adding or removing a NIC means (1) running nixos-generate-config only to see it updates hardware-configuration.nix and not configuration.nix, then having to (2) manually copy over the entries.

>if we fix this issue, by moving the generated stuff into hardware-configuration.nix and use mkDefault

Yes, that would be a better workaround for #75515. Best would be to fix `networking.useDHCP` instead of deprecating it.

>When you put `networking.interfaces.enp0s31f6.useDHCP = true;` into hardware-configuration.nix, you can't change it to false

When `mkDefault` is used, my concern no longer applies.
